### PR TITLE
[v7r0] Modified flag for using the JobParameters on ES

### PR DESCRIPTION
**Pull request by fstagni**
BEGINRELEASENOTES

*WMS
CHANGE: Modified flag for using the JobParameters on ES

*docs
NEW: added a few notes on using the JobParameters on ES

*test
NEW: integration test for JobParameters on ES

ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
Please review!

File: Core/Utilities/ElasticSearchDB.py [hunk 1]
```diff
@@ -251,21 +251,23 @@ def deleteIndex(self, indexName):
...
       res = self.__client.index(index=indexName,
                                 doc_type=doc_type,
-                                body=body)
+                                body=body,
+                                id=id)
     except TransportError as e:
       return S_ERROR(e)
 
-    if res.get('created') or res.get('result') == 'created':
+    if res.get('created') or res.get('result') == 'created' or res.get('result') == 'updated':
```
Comment (chrisburr) [inline Core/Utilities/ElasticSearchDB.py]:
```suggestion
    if res.get('created') or res.get('result') in ('created', 'updated'):
```

File: WorkloadManagementSystem/DB/ElasticJobDB.py [hunk 2]
```diff
@@ -80,151 +74,24 @@ def getJobParameters(self, jobID, paramList=None):
       except BaseException:
```
Comment (chrisburr) [inline WorkloadManagementSystem/DB/ElasticJobDB.py]:
```suggestion
      except Exception:
```

File: WorkloadManagementSystem/Service/JobMonitoringHandler.py [hunk 3]
```diff
@@ -515,7 +507,11 @@ def export_getJobParameter(jobID, parName):
     :param str parName: one single parameter name
     """
     if gElasticJobDB:
-      return gElasticJobDB.getJobParameters(jobID, [parName])
+      res = gElasticJobDB.getJobParameters(jobID, [parName])
+      if not res['OK']:
+        return res
+      if res['Value'].get(int(jobID)):
+        return S_OK(res['Value'][jobID])
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/JobMonitoringHandler.py]:
```suggestion
        return S_OK(res['Value'][int(jobID)])
```
If there is a cast in the get before, there needs to be a cast here

File: WorkloadManagementSystem/Service/JobMonitoringHandler.py [hunk 4]
```diff
@@ -539,7 +535,30 @@ def export_getJobParameters(jobIDs, parName=None):
...
+      res = gJobDB.getJobParameters(jobIDs, parName)
+      if not res['OK']:
+        return res
+      parametersM = res['Value']
+
+      # and now combine
+      final = dict(parametersM)
+      for jobID in parametersM:
+        final[jobID].update(parameters.get(jobID, {}))
+      for jobID in parameters.iterkeys():
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/JobMonitoringHandler.py]: Why `.iterkeys()`? and not just `for jobID in parameters:` ?

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py [hunk 5]
```diff
@@ -48,15 +49,13 @@ def initialize(self):
 
     Determines the switching of ElasticSearch and MySQL backends
     """
-    global elasticJobDB, jobDB
+    global elasticJobDB
 
-    gESFlag = self.srv_getCSOption('useES', False)
-    if gESFlag:
+    useESForJobParametersFlag = Operations().getValue('/Services/JobMonitoring/useESForJobParametersFlag', False)
+    global elasticJobDB
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/JobStateUpdateHandler.py]: one `global elasticJobDB` is enough, isn't it? (also in line 52)

File: WorkloadManagementSystem/Service/SandboxStoreHandler.py [hunk 6]
```diff
@@ -487,6 +479,14 @@ def __deleteSandboxFromExternalBackend(self, SEName, SEPFN):
     if self.getCSOption("DelayedExternalDeletion", True):
       gLogger.info("Setting deletion request")
       try:
+
+        # We need the hostDN used in order to pass these credentials to the
+        # SandboxStoreDB..
+        hostCertLocation, _ = Locations.getHostCertificateAndKeyLocation()
+        hostCert = X509Certificate.X509Certificate()
+        hostCert.loadFromFile(hostCertLocation)
+        self.hostDN = hostCert.getSubjectDN().get('Value')
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/SandboxStoreHandler.py]: No need to keep this as an class attribute, so drop the `self.`?

**Comment by fstagni** [comment]
Fixed both suggestions.

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py [hunk 7]
```diff
@@ -330,10 +326,18 @@ def export_setJobsParameter(self, jobsParameterDict):
...
       if elasticJobDB:
-        elasticJobDB.setJobParameter(jobID, str(jobsParameterDict[jobID][0]), str(jobsParameterDict[jobID][1]))
-
-      if jobDB:
-        jobDB.setJobParameter(jobID, str(jobsParameterDict[jobID][0]), str(jobsParameterDict[jobID][1]))
+        res = elasticJobDB.setJobParameter(jobID,
+                                           str(jobsParameterDict[jobID][0]),
+                                           str(jobsParameterDict[jobID][1]))
+        if not res['OK']:
+          self.log.error(res['Message'])
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/JobStateUpdateHandler.py]:
```suggestion
          self.log.error('Failed to add Elastic Job Parameter', res['Message'])
```

File: WorkloadManagementSystem/Service/JobStateUpdateHandler.py [hunk 8]
```diff
@@ -330,10 +326,18 @@ def export_setJobsParameter(self, jobsParameterDict):
...
+
+      else:
+        res = jobDB.setJobParameter(jobID,
+                                    str(jobsParameterDict[jobID][0]),
+                                    str(jobsParameterDict[jobID][1]))
+        if not res['OK']:
+          self.log.error(res['Message'])
```
Comment (andresailer) [inline WorkloadManagementSystem/Service/JobStateUpdateHandler.py]:
```suggestion
          self.log.error('Failed to add MySQL Job Parameter', res['Message'])
```

**Comment by fstagni** [comment]
Thanks @andresailer , updated

**Pull request merged**